### PR TITLE
Edit and update the glossary

### PR DESCRIPTION
**Pull request by gguillotte**
The glossary has some typos and outdated information, and is missing some commonly used terms (like Puppet run, puppetize, and several product names).

### CODE REVIEW

File: source/references/glossary.markdown [hunk 1]
```diff
@@ -1,82 +1,127 @@
 ---
-title: Glossary
+title: Glossary of Puppet Vocabulary
 layout: default
+canonical: "/references/glossary.html"
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: If the canonical link is itself, and it's not a versioned page that'll keep moving forward, then you don't need to add an explicit entry; we automatically set it to self.

File: source/references/glossary.markdown [hunk 2]
```diff
@@ -1,82 +1,127 @@
...
+You specify an attribute's value with the `=>` operator, and pairs of attributes and values are separated by commas.
 
 ### agent
 
-(or  **agent node**)
+(or **Puppet agent**, or **agent node**)
+
+"**Puppet agent**" can refer to a few different things:
+
+* **The Puppet agent software package.** Puppet is usually deployed in a simple client-server arrangement. When you install Puppet on a device, you're installing the agent software and its associated tools and dependencies. On *nix systems, this package is called `puppet-agent`.
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: It's called puppet-agent on windows now too!

File: source/references/glossary.markdown [hunk 3]
```diff
@@ -404,33 +594,50 @@ A kind of [resource](#resource) that Puppet is able to manage; for example, `fil
...
 Puppet ships with a set of built-in resource types; see the [type reference](/references/stable/type.html) for a complete list of them. New [native types](#type-native) can be added as [plugins](#plugin), and [defined types](#type-defined) can be constructed by grouping together resources of existing types.
 
+Contrast with [data type](#type-data). See also [defined type](#type-defined) and [native type](#type-native).
+
+### type (data)
+
+(or **data type**; includes **abstract data type** and **core data type**)
+
+Every [value](#value) has a **data type**, which is a named classification of a type of data that a [variable](#variable) or [parameter](#parameter) can hold. The [Puppet language](#puppet-language) has core data types (such as integer, Boolean, or string) and abstract data types (such as any or optional). For more information, see [About Values and Data Types](/puppet/latest/reference/lang_data.html) in the Puppet language reference.
+
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: :+1: to this section.

File: source/references/glossary.markdown [hunk 4]
```diff
@@ -251,148 +358,231 @@ Generally speaking, a parameter is a chunk of information that a class or resour
...
+A **site** refers to an entire IT ecosystem that is managed by Puppet. A site includes all Puppet [master](#master) servers, [agent nodes](#agent), and independent masterless Puppet [nodes](#node) within an organization.
 
 ### site module
 
-A common idiom in which one or more [modules](#module) contain [classes](#class) specific to a given Puppet site. These classes usually describe complete configurations for a specific system or a given group of systems. For example, the `site::db_slave` class might describe the entire configuration of a database server, and a new database server could be configured simply by applying that class to it.
+A **site module** is a common Puppet idiom in which one or more [modules](#module) contain [classes](#class) specific to a given Puppet [site](#site). These classes usually describe complete configurations for a specific system or group of systems. For example, the `site::db_slave` class might describe the entire configuration of a database server, and a new database server could be configured simply by applying that class to it.
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: mmmmmmmmm this approach was kind of obsoleted by the roles and profiles pattern. ...which we still don't have great docs for.

File: source/references/glossary.markdown [hunk 5]
```diff
@@ -84,166 +129,228 @@ See [external node classifier](#external-node-classifier).
...
-The puppet master uses an HTTP server to provide catalogs. It can run as a standalone daemon process with a built-in web server, or it can be managed by a production-grade web server that supports the rack API. The built-in web server is meant for testing, and is not suitable for use with more than ten nodes.
+The Puppet master provides catalogs to agents using an HTTP server. It can run as a standalone daemon with a built-in web server, or as part of [Puppet Server](#puppet-server). (We don't recommend using the built-in daemon when managing more than 10 nodes.)
+
+For more information, see the [Overview of Puppet's Architecture](/puppet/latest/reference/architecture.html) page in the Puppet documentation.
+
+### masterless
+
+"**Masterless**" is a term often used to refer to a Puppet [node](#node) that does not rely on a [master](#master) for its [catalog](#catalog). See [agent](#agent).
 
 ### manifest
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: ...This is out of alphabetical order.

File: source/references/glossary.markdown [hunk 6]
```diff
@@ -84,166 +129,228 @@ See [external node classifier](#external-node-classifier).
...
+See [scope](#scope).
 
-### inheritance (node)
+### main manifest
 
-Node statements can be derived from other node statements with the `inherits` keyword. This works identically to the way class inheritance works.
+(or **site manifest**)
 
-> **Note:** Node inheritance **should almost always be avoided.** Many new users attempt to use node inheritance to look up variables that have a common default value and a rare specific value on certain nodes; it is not suited to this task, and often yields the opposite of the expected result. If you have a lot of conditional per-node data, we recommend using the Heira tool or assigning variables with an ENC instead.
+The **main manifest** is the main "point of entry" [manifest](#manifest) used by a Puppet [master](#master) when [compiling](#compilation) a [catalog](#catalog). A main manifest specific to a [site](#site) is called a **site manifest**.
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: Mmm, it's more like "site manifest" was just an outdated name for "main manifest."

File: source/references/glossary.markdown [hunk 7]
```diff
@@ -84,166 +129,228 @@ See [external node classifier](#external-node-classifier).
...
 
-Most manifests are contained in [modules](#module). Every manifest in a module should [define](#define) a single [class](#class) or [defined type](#type-defined).
+The Puppet [master](#master) service reads an [environment's](#environment) [main manifest](#main-manifest). This manifest usually defines [nodes](#node-definition), so that each managed [agent](#agent) receives a unique [catalog](#catalog).
 
-The puppet master service reads a single "site manifest," usually located at `/etc/puppet/manifests/site.pp`. This manifest usually defines [nodes](#node-definition), so that each managed [agent node](#node) will receive a unique catalog.
+See also [main manifest](#main-manifest).
+
+### manifest ordering
+
+Puppet uses **manifest ordering** to apply [resources](#resource) in the order that they're declared in their [manifests](#manifest). For more information, see the [Relationships and Ordering][] page in the [Puppet language](#puppet-language) reference.
```
Comment (nfagerlund) [inline source/references/glossary.markdown]: I think it's worth saying that this is for unrelated resources, and that related resources should have explicit relationships.

File: source/references/glossary.markdown (hunk 1 above)
Comment (gguillotte) [inline source/references/glossary.markdown]: Do we ever want to consider versioning the glossary? Or is it a bad fit for the glossary to describe how term has different or new meanings across versions?

File: source/references/glossary.markdown (hunk 4 above)
Comment (gguillotte) [inline source/references/glossary.markdown]: "Site module" shows up enough in the current docs (particularly in the PE and modules docs) that we should still define it. And for what it's worth, this PR includes new entries for roles and profiles that link to the [PE docs' description](https://docs.puppetlabs.com/pe/latest/puppet_assign_configurations.html#assigning-configuration-data-with-role-and-profile-modules).

We should revisit whether to include this once we have satisfactory roles and profiles docs.

File: source/references/glossary.markdown (hunk 6 above)
Comment (gguillotte) [inline source/references/glossary.markdown]: OK. We should scrub the docs to use or prioritize main manifest over site manifest.

**Comment by gguillotte** [comment]
Amended commit with some of the fixes suggested by @nfagerlund

**Pull request merged**